File: src/main/java/plazma/kernel/lib/fmt/FmtLib.java
/*
 * Copyright (C) 2012-2023 Oleh Hapon ohapon@users.sourceforge.net
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307, USA.
 * 
 * Oleh Hapon
 * Kyiv, UKRAINE
 * ohapon@users.sourceforge.net
 */

package plazma.kernel.lib.fmt;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

import plazma.kernel.lib.str.StrLib;

public class FmtLib {

    // Functions:

    /////////////////////////////////////////////////////////////////////////////////
    //
    // Common
    //
    // 1.1
    //
    // - String format(String message, Object... args)
    //

    /////////////////////////////////////////////////////////////////////////////////
    //
    // Date/Time/DateTime
    //
    // 2.1
    //
    // - String formatDate()
    // - String formatDate(String format)
    // - String formatDate(DateFormat formatter)
    //
    // - String formatDate(Date date)
    // - String formatDate(Date date, String format)
    // - String formatDate(Date date, DateFormat formatter)
    //
    // 2.2
    //
    // - String formatTime()
    // - String formatTime(String format)
    // - String formatTime(DateFormat formatter)
    //
    // - String formatTime(Date date)
    // - String formatTime(Date date, String format)
    // - String formatTime(Date date, DateFormat formatter)
    //
    // 2.3
    //
    // - String formatDateTime()
    // - String formatDateTime(String format)
    // - String formatDateTime(DateFormat formatter)
    //
    // - String formatDateTime(Date date)
    // - String formatDateTime(Date date, String format)
    // - String formatDateTime(Date date, DateFormat formatter)

    /////////////////////////////////////////////////////////////////////////////////
    //
    // Date/Time/DateTimeFormatter
    //
    // 3.1
    //
    // - DateFormat getDateFormatter(String format)
    // - DateFormat getTimeFormatter(String format)
    // - DateFormat getDateTimeFormatter(String format)

    public static final String DEFAULT_DATE_FORMAT = "yyyy-MM-dd";
    public static final String DEFAULT_TIME_FORMAT = "HH:mm:ss";
    public static final String DEFAULT_DATE_TIME_FORMAT = "yyyy-MM-dd HH:mm:ss";

    public static final String EMPTY_FORMAT_VALUE = "";
    private static final boolean USE_NORMALIZE_FORMAT = true;

    private static final Map<String, String> dateFormats = new HashMap<>();

    static {

        ////////////////////////////////////////////////////////////////
        //
        // Date
        //
        ////////////////////////////////////////////////////////////////

        // ISO 8601
        dateFormats.put("YYYY-MM-DD", "yyyy-MM-dd");
        dateFormats.put("YYYY.MM.DD", "yyyy.MM.dd");
        dateFormats.put("DD.MM.YYYY", "dd.MM.yyyy");
        dateFormats.put("DD/MM/YYYY", "dd/MM/yyyy");
        dateFormats.put("MM/DD/YYYY", "MM/dd/yyyy");

        // Unicode
        dateFormats.put("yyyy-MM-dd", "yyyy-MM-dd");
        dateFormats.put("yyyy.MM.dd", "yyyy.MM.dd");
        dateFormats.put("dd.MM.yyyy", "dd.MM.yyyy");
        dateFormats.put("dd/MM/yyyy", "dd/MM/yyyy");
        dateFormats.put("MM/dd/yyyy", "MM/dd/yyyy");

        ////////////////////////////////////////////////////////////////
        //
        // Time
        //
        ////////////////////////////////////////////////////////////////

        // ISO 8601
        dateFormats.put("hh:mm:ss", "HH:mm:ss");

        // Unicode
        dateFormats.put("HH:mm:ss", "HH:mm:ss");

        ////////////////////////////////////////////////////////////////
        //
        // DateTime
        //
        ////////////////////////////////////////////////////////////////

        // ISO 8601
        dateFormats.put("YYYY-MM-DD hh:mm:ss", "yyyy-MM-dd HH:mm:ss");
        dateFormats.put("YYYY.MM.DD hh:mm:ss", "yyyy.MM.dd HH:mm:ss");
        dateFormats.put("DD.MM.YYYY hh:mm:ss", "dd.MM.yyyy HH:mm:ss");
        dateFormats.put("DD/MM/YYYY hh:mm:ss", "dd/MM/yyyy HH:mm:ss");
        dateFormats.put("MM/DD/YYYY hh:mm:ss", "MM/dd/yyyy HH:mm:ss");

        // Unicode
        dateFormats.put("yyyy-MM-dd HH:mm:ss", "yyyy-MM-dd HH:mm:ss");
        dateFormats.put("yyyy.MM.dd HH:mm:ss", "yyyy.MM.dd HH:mm:ss");
        dateFormats.put("dd.MM.yyyy HH:mm:ss", "dd.MM.yyyy HH:mm:ss");
        dateFormats.put("dd/MM/yyyy HH:mm:ss", "dd/MM/yyyy HH:mm:ss");
        dateFormats.put("MM/dd/yyyy HH:mm:ss", "MM/dd/yyyy HH:mm:ss");

    }

    private FmtLib() {
    }

    private static String _normalizeFormat(Map<String, String> formats, String format, String defaultFormat) {

        // Normalize string
        format = normalize(format);
        if (isEmpty(format)) {
            return defaultFormat;
        }

        if (!USE_NORMALIZE_FORMAT) {
            return format;
        }

        // Normalize format
        String normalizeFormat = formats.get(format);
        if (normalizeFormat != null) {
            return normalizeFormat;
        }
        return format;
    }

    //// 1.1

    public static final String format(String message, Object... args) {
        return String.format(message, args);
    }

    //// 2.1 formatDate

    public static String formatDate() {
        return formatDate(_getCurrentDate());
    }

    public static String formatDate(String format) {
        return formatDate(_getCurrentDate(), format);
    }

    public static String formatDate(DateFormat formatter) {
        return formatDate(_getCurrentDate(), formatter);
    }

    //

    public static String formatDate(Date date) {
        return _formatDate(date, (String) null, DEFAULT_DATE_FORMAT);
    }

    public static String formatDate(Date date, String format) {
        return _formatDate(date, format, DEFAULT_DATE_FORMAT);
    }

    public static String formatDate(Date date, DateFormat formatter) {
        return _formatDate(date, formatter, DEFAULT_DATE_FORMAT);
    }

    //// 2.2 formatTime

    public static String formatTime() {
        return formatTime(_getCurrentDate());
    }

    public static String formatTime(String format) {
        return formatTime(_getCurrentDate(), format);
    }

    public static String formatTime(DateFormat formatter) {
        return formatTime(_getCurrentDate(), formatter);
    }

    //

    public static String formatTime(Date date) {
        return _formatDate(date, (String) null, DEFAULT_TIME_FORMAT);
    }

    public static String formatTime(Date date, String format) {
        return _formatDate(date, format, DEFAULT_TIME_FORMAT);
    }

    public static String formatTime(Date date, DateFormat formatter) {
        return _formatDate(date, formatter, DEFAULT_TIME_FORMAT);
    }

    //// 2.3 formatDateTime

    public static String formatDateTime() {
        return formatDateTime(_getCurrentDate());
    }

    public static String formatDateTime(String format) {
        return formatDateTime(_getCurrentDate(), format);
    }

    public static String formatDateTime(DateFormat formatter) {
        return formatDateTime(_getCurrentDate(), formatter);
    }

    //

    public static String formatDateTime(Date date) {
        return _formatDate(date, (String) null, DEFAULT_DATE_TIME_FORMAT);
    }

    public static String formatDateTime(Date date, String format) {
        return _formatDate(date, format, DEFAULT_DATE_TIME_FORMAT);
    }

    public static String formatDateTime(Date date, DateFormat formatter) {
        return _formatDate(date, formatter, DEFAULT_DATE_TIME_FORMAT);
    }

    //// 3.1

    public static DateFormat getDateFormatter(String format) {
        return _getDateFormatter(format, DEFAULT_DATE_FORMAT);
    }

    public static DateFormat getTimeFormatter(String format) {
        return _getDateFormatter(format, DEFAULT_TIME_FORMAT);
    }

    public static DateFormat getDateTimeFormatter(String format) {
        return _getDateFormatter(format, DEFAULT_DATE_TIME_FORMAT);
    }

    ////

    private static DateFormat _getDateFormatter(String format, String defaulFormat) {
        format = _normalizeFormat(dateFormats, format, defaulFormat);
        return format == null ? null : new SimpleDateFormat(format);
    }

    //

    private static String _formatDate(Date date, DateFormat formatter, String defaultFormat) {
        if (date == null) {
            return EMPTY_FORMAT_VALUE;
        }

        // Normalize formatter
        if (formatter == null && defaultFormat != null) {
            formatter = _getDateFormatter(defaultFormat, null);
        }
        return _formatDate(date, formatter);
    }

    private static String _formatDate(Date date, String format, String defaultFormat) {
        if (date == null) {
            return EMPTY_FORMAT_VALUE;
        }
        DateFormat formatter = _getDateFormatter(format, defaultFormat);
        return _formatDate(date, formatter);
    }

    private static String _formatDate(Date date, DateFormat formatter) {
        if (date == null || formatter == null) {
            return EMPTY_FORMAT_VALUE;
        }
        return formatter.format(date);
    }

    //

    private static Date _getCurrentDate() {
        return new Date();
    }

    private static String normalize(String str) {
        return StrLib.normalize(str);
    }

    private static boolean isEmpty(String str) {
        return StrLib.isEmpty(str);
    }

}

File: src/main/cpp/plazma/kernel/lib/fmt/fmtlib.cpp
#include <string>
#include <chrono>

#include <vector>
#include <iterator>
#include <map>

#include "fmtlib.h"

/////////////////////////////////////////////////////////////////////////////////
// 
// Common
// 
// 1.1
//
// - string format(const string& message, const string& value)
//
    
/////////////////////////////////////////////////////////////////////////////////
//
// Date/Time/DateTime
//
// 2.1
//
// - string formatDate()
// - string formatDate(string format)
//
// - string formatDate(time_point date)
// - string formatDate(time_point date, string format)
//
// 2.2
//
// - string formatTime()
// - string formatTime(string format)
//
// - string formatTime(time_point date)
// - string formatTime(time_point date, string format)
//
// 2.3
//
// - string formatDateTime()
// - string formatDateTime(time_point format)
//
// - string formatDateTime(time_point date)
// - string formatDateTime(time_point date, string format)
//

namespace fmtlib {

const std::string DEFAULT_DATE_FORMAT = "%Y-%m-%d";
const std::string DEFAULT_TIME_FORMAT = "%H:%M:%S";
const std::string DEFAULT_DATE_TIME_FORMAT = "%Y-%m-%d %H:%M:%S";

const std::string EMPTY_FORMAT_VALUE = "";
const bool USE_NORMALIZE_FORMAT = true;

std::map<std::string, std::string> dateFormats;
bool initFlag = false;

void _init() {

    initFlag = true;

    ////////////////////////////////////////////////////////////////
	//
	// Date
	//
	////////////////////////////////////////////////////////////////

	// ISO 8601
	dateFormats["YYYY-MM-DD"] = "%Y-%m-%d";
	dateFormats["YYYY.MM.DD"] = "%Y.%m.%d";
	dateFormats["DD.MM.YYYY"] = "%d.%m.%Y";
	dateFormats["DD/MM/YYYY"] = "%d/%m/%Y";
	dateFormats["MM/DD/YYYY"] = "%m/%d/%Y";

	// Unicode
	dateFormats["yyyy-MM-dd"] = "%Y-%m-%d";
	dateFormats["yyyy.MM.dd"] = "%Y.%m.%d";
	dateFormats["dd.MM.yyyy"] = "%d.%m.%Y";
	dateFormats["dd/MM/yyyy"] = "%d/%m/%Y";	
    dateFormats["MM/dd/yyyy"] = "%m/%d/%Y";



    ////////////////////////////////////////////////////////////////
	//
	// Time
	//
	////////////////////////////////////////////////////////////////
	
	// ISO 8601
	dateFormats["hh:mm:ss"] = "%H:%M:%S";
	
	// Unicode
	dateFormats["HH:mm:ss"] = "%H:%M:%S";	



	////////////////////////////////////////////////////////////////
	//
	// DateTime
	//
	////////////////////////////////////////////////////////////////
	
	// ISO 8601	
	dateFormats["YYYY-MM-DD hh:mm:ss"] = "%Y-%m-%d %H:%M:%S";
	dateFormats["YYYY.MM.DD hh:mm:ss"] = "%Y.%m.%d %H:%M:%S";
	dateFormats["DD.MM.YYYY hh:mm:ss"] = "%d.%m.%Y %H:%M:%S";
	dateFormats["DD/MM/YYYY hh:mm:ss"] = "%d/%m/%Y %H:%M:%S";
	dateFormats["MM/DD/YYYY hh:mm:ss"] = "%m/%d/%Y %H:%M:%S";

	// Unicode
	dateFormats["yyyy-MM-dd HH:mm:ss"] = "%Y-%m-%d %H:%M:%S";
	dateFormats["yyyy.MM.dd HH:mm:ss"] = "%Y.%m.%d %H:%M:%S";
	dateFormats["dd.MM.yyyy HH:mm:ss"] = "%d.%m.%Y %H:%M:%S";
	dateFormats["dd/MM/yyyy HH:mm:ss"] = "%d/%m/%Y %H:%M:%S";
	dateFormats["MM/dd/yyyy HH:mm:ss"] = "%m/%d/%Y %H:%M:%S";

}

// 1.1 Common

std::string _vformat(/*FILE *fp,*/ const char *fmt, va_list va) {

    // https://stackoverflow.com/questions/1657883/variable-number-of-arguments-in-c
    // https://stackoverflow.com/questions/7315936/which-of-sprintf-snprintf-is-more-secure
    // https://en.cppreference.com/w/cpp/utility/variadic/va_arg

    va_list va2;
    va_copy(va2, va);

    size_t n = std::vsnprintf(NULL, 0, fmt, va2) + 1; // +1 for the '\0'
 
    char str[n]; // 50
    //int n = sizeof str;

    //const char* v1 = va_arg(va, const char*);
    std::vsnprintf(str, n, fmt, va);

    return str;
}

std::string vformat(const char *fmt, ...) {
    va_list va;
    va_start(va, fmt);
    std::string str = _vformat(fmt, va);
    va_end(va);
    return str;
}

std::string format(const std::string& message, const std::string& value) {
	return vformat(message.c_str(), value.c_str());
    //char buff[100]; // TODO: what about dynamic size?
    //snprintf(buff, sizeof(buff), message.c_str(), value.c_str());
    //std::string result = buff;
    //return result;
}

std::string format(const std::string& message, const std::string& value1, const std::string& value2) {
	return vformat(message.c_str(), value1.c_str(), value2.c_str());
}

////

std::string _normalizeFormat(std::map<std::string, std::string>& formats, const std::string& format, const std::string& defaultFormat) {
	
	// Normalize string
	//format = normalize(format);

	std::string result;
	if (format.empty()) {
		result = defaultFormat;
	    return result;
	}
	
	if (!USE_NORMALIZE_FORMAT) {
		result = format;
	    return result;
	}
	
	// Normalize format
	result = formats[format];
	if (!result.empty()) {
	    return result;
	}
	result = format;
	return result;
}

std::string _formatDate(const std::chrono::system_clock::time_point& date, const std::string& format, const std::string& defaultFormat) {

    // Initialization
    if (!initFlag) {
        _init();
    }

    // Normalize format
    std::string resultFormat = _normalizeFormat(dateFormats, format, defaultFormat);

    std::time_t now = std::chrono::system_clock::to_time_t(date);

    std::string result(30, '\0'); // TODO: what about dynamic size?
    std::strftime(&result[0], result.size(), resultFormat.c_str(), std::localtime(&now));

    return result;
}

std::string _formatDate(const std::chrono::system_clock::time_point& date, const std::string& format) {
    return _formatDate(date, format, "");
}

std::chrono::system_clock::time_point _getCurrentDate() {
    return std::chrono::system_clock::now();
}

// 2.1 Date

std::string formatDate() {
    return _formatDate(_getCurrentDate(), DEFAULT_DATE_FORMAT);
}

std::string formatDate(const std::string& format) {
    return _formatDate(_getCurrentDate(), format, DEFAULT_DATE_FORMAT);
}

//

std::string formatDate(const std::chrono::system_clock::time_point& date) {
    return _formatDate(date, DEFAULT_DATE_FORMAT);
}

std::string formatDate(const std::chrono::system_clock::time_point& date, const std::string& format) {
    return _formatDate(date, format, DEFAULT_DATE_FORMAT);
}

// 2.2 Time

std::string formatTime() {
    return _formatDate(_getCurrentDate(), DEFAULT_TIME_FORMAT);
}

std::string formatTime(const std::string& format) {
    return _formatDate(_getCurrentDate(), format, DEFAULT_TIME_FORMAT);
}

//

std::string formatTime(const std::chrono::system_clock::time_point& date) {
    return _formatDate(date, DEFAULT_TIME_FORMAT);
}

std::string formatTime(const std::chrono::system_clock::time_point& date, const std::string& format) {
    return _formatDate(date, format, DEFAULT_TIME_FORMAT);
}

// 2.3 DateTime

std::string formatDateTime() {
    return _formatDate(_getCurrentDate(), DEFAULT_DATE_TIME_FORMAT);
}

std::string formatDateTime(const std::string& format) {
    return _formatDate(_getCurrentDate(), format, DEFAULT_DATE_TIME_FORMAT);
}

//

std::string formatDateTime(const std::chrono::system_clock::time_point& date) {
    return _formatDate(date, DEFAULT_DATE_TIME_FORMAT);
}

std::string formatDateTime(const std::chrono::system_clock::time_point& date, const std::string& format) {
    return _formatDate(date, format, DEFAULT_DATE_TIME_FORMAT);
}

}